pkgs/top-level/all-packages.nix

```diff
@@ -1354,6 +1354,11 @@ in
 
   bs-platform = callPackage ../development/compilers/bs-platform {};
 
+  reason-language-server = callPackage ../development/tools/reason-language-server {
+    ocamlPackages = ocaml-ng.ocamlPackages_4_08;
+    inherit (ocaml-ng.ocamlPackages_4_08) buildDunePackage;
+  };
+
```

Comment:
Maybe this should live in `pkgs/top-level/ocaml-packages.nix`? I couldn't find a nice way to force OCaml 4.08 when it was there though

Reply:
Agree it should.

You can do something like this: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/ocaml/dune.nix#L7-L10

And it would avoid having to `override` `ocamlPackages` in the reason derivation.

Reply:
The problem is that there is no `maximumOCamlVersion` argument anywhere - `reason-language-server` works with OCaml versions before 4.09. I don't know the ocaml infrastructure well at all (this is my first look at it), so I would really appreciate it if someone more knowledgeable could maybe add a commit to this pr if this is something that requires changes.

Reply:
What about changing the expression so that instead of `buildDunePackage` and `ocamlPackages` it takes `ocaml-ng` and then picking these two things within the `pkgs/development/tools/reason-language-server/default.nix` ?

Reply:
Something else I remembered: given that Reason is a build time only dependency, does it even need to be on the same OCaml version than RLS? AFAIK there just needs to be a `refmt` binary in `$PATH` for Dune to compile a Reason project.

Reply:
> Something else I remembered: given that Reason is a build time only dependency, does it even need to be on the same OCaml version than RLS? AFAIK there just needs to be a `refmt` binary in `$PATH` for Dune to compile a Reason project.

While I was packaging it seemed to break unless Reason used the same OCaml version.

> What about changing the expression so that instead of `buildDunePackage` and `ocamlPackages` it takes `ocaml-ng` and then picking these two things within the `pkgs/development/tools/reason-language-server/default.nix` ?

I don't have strong feelings either way, but I feel like someone who uses OCaml a lot more than me should make the decision.

Reply:
Does the user of reason-language-server care about the version of OCaml that is used to build it? (i.e., does RLS behave differently depending on the version of OCaml used to build it?) Programs that use native OCaml serialization often do (for instance: unison, Coq, merlin).

If this is the case, it may make sense to put it inside `ocamlPackages`.

Otherwise just take `ocamlPackages` as argument (don’t take neither `reason` nor `buildDunePackage` as argument: read them from `ocamlPackages`).

In either case, `reason` should not take the whole `ocamlPackage` set as argument (because it is part of it).

Reply:
yup exactly my thoughts @vbgl :+1:

Reply:
I think there's an argument that `reason` shouldn't be a part of `ocamlPackages` to be made - seems like the use case of compiling `reason` against different sets of `ocamlPackages` is going to be a lot more common than overriding specific dependencies, and it is supposed to be its own whole ecosystem.

> Does the user of reason-language-server care about the version of OCaml that is used to build it? (i.e., does RLS behave differently depending on the version of OCaml used to build it?) Programs that use native OCaml serialization often do (for instance: unison, Coq, merlin).

I'm not sure of the answer to this question. Someone who has used reason more should weigh in.

Reply:
Whatever the answer to this question, the commit “reason: use the whole ocamlPackages set as a variable” is wrong.

In particular, its explanation (“This makes it much easier to override the version of ocaml reason is built against (`reason.override { ocamlPackages = ...; }` instead of having to override `ocaml`, `findlib`, `dune`, etc.”) is nonsense. If you want a `reason` built with a specific set of OCaml stuff, just use `ocaml-ng.ocamlPackages_XXX.reason`.